pkg/evs/controllerserver.go

```diff
@@ -266,18 +269,181 @@ func (cs *ControllerServer) ListVolumes(_ context.Context, req *csi.ListVolumesR
 
 func (cs *ControllerServer) CreateSnapshot(_ context.Context, req *csi.CreateSnapshotRequest) (
 	*csi.CreateSnapshotResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+	klog.Infof("CreateSnapshot called with request %v", *req)
+	credentials := cs.Driver.cloudCredentials
+	name := req.GetName()
+	volumeId := req.GetSourceVolumeId()
+	if err := checkCreateSnapshotParamEmpty(name, volumeId); err != nil {
+		klog.Errorf("Failed to check create snapshot param, %v", err)
+		return nil, err
+	}
+	response, err := checkDuplicateSnapshotName(credentials, name, volumeId)
+	if err != nil {
+		klog.Errorf("Failed to check duplicate snapshot name: %v", err)
+		return nil, err
+	}
+	if response != nil {
+		klog.Infof("Snapshot with name: %s | volumeId: %s already exist. detail: %v", name, volumeId, response)
+		return response, nil
+	}
+	if err := checkVolumeIsExist(credentials, volumeId); err != nil {
+		klog.Errorf("Failed to check volume is exist. volumeId: %s, error: %v", volumeId, err)
+		return nil, err
+	}
+	response, err = services.CreateSnapshotToCompleted(credentials, name, volumeId)
+	if err != nil {
+		klog.Errorf("Failed to create snapshot to completed: %v", err)
+		return nil, err
+	}
+	klog.Infof("Successful create snapshot. detail: %v", response)
+	return response, nil
+}
+
+func checkCreateSnapshotParamEmpty(name string, volumeId string) error {
+	if volumeId == "" {
+		return status.Error(codes.InvalidArgument, "CreateSnapshot volumeId cannot be empty")
+	}
+	if name == "" {
+		return status.Error(codes.InvalidArgument, "CreateSnapshot name cannot be empty")
+	}
+	return nil
+}
+
+func checkDuplicateSnapshotName(credentials *config.CloudCredentials, name string, volumeId string) (
+	*csi.CreateSnapshotResponse, error) {
+	listOpts := snapshots.ListOpts{
+		Name: name,
+	}
+	listSnapshots, err := services.ListSnapshots(credentials, listOpts)
+	if err != nil {
+		return nil, err
+	}
+	if len(listSnapshots) == 1 {
+		snap := &listSnapshots[0]
+		if snap.VolumeID != volumeId {
+			return nil, status.Error(codes.AlreadyExists, "CreateSnapshot same name with different volumeId")
+		}
+		return &csi.CreateSnapshotResponse{
+			Snapshot: &csi.Snapshot{
+				SnapshotId:     snap.ID,
+				SizeBytes:      int64(snap.Size * common.GbByteSize),
+				SourceVolumeId: snap.VolumeID,
+				CreationTime:   timestamppb.New(snap.CreatedAt),
+				ReadyToUse:     true,
+			},
+		}, nil
+	}
+	if len(listSnapshots) > 1 {
+		return nil, status.Error(codes.Internal, "Multiple snapshots reported by EVS with same name")
+	}
+	return nil, nil
+}
+
+func checkVolumeIsExist(credentials *config.CloudCredentials, volumeId string) error {
+	volume, err := services.GetVolume(credentials, volumeId)
+	if err != nil {
+		return err
```

Comment:
the error cannot be treated as if the volume already exists
when it does not exist, it will return a 404 error, which needs to be checked with common.IsNotFound